### PR TITLE
Reorganize training scripts and add LoRA comparison

### PR DESCRIPTION
**Pull request by pinecypressfxd**
Updated Python to the latest version and all dependent packages (0813).
Added LoRA fine-tuning, calling LoRA via function parameters.
Added loss visualization and compared LoRA with full fine tuning.
Reorganize training scripts and add LoRA comparison, revise the parameter type, add assert.

更新python版本到最新版本，同步更新各个依赖包到最新版（0806），参见requirements。
增加LoRA微调，通过函数参数调用LoRA。
增加loss的可视化，并比对LoRA和full fine tune。
重新组织训练脚本名称，修改依赖包，bool类型，增加assert等细节。

### CODE REVIEW

File: finetune_distributed.py [hunk 1]
```diff
@@ -242,26 +481,64 @@ def train():
...
+            state_dict=accelerator.get_state_dict(model),
+        )
+    
     if accelerator.is_local_main_process:
         processor.save_pretrained(output_dir)
         write_chat_template(processor, output_dir)
 
+    # Plot loss history
+    if accelerator.is_local_main_process:
+        training_type = "LoRA" if use_lora and hasattr(model, 'peft_config') else "Full"
```
Comment (zhangfaen) [inline finetune_distributed.py]: “and hasattr(model, 'peft_config')”  这部分去掉？

File: finetune_distributed.py [hunk 2]
```diff
@@ -242,26 +481,64 @@ def train():
...
+
 if __name__ == "__main__":
-    train()
+    import argparse
+    parser = argparse.ArgumentParser(description='Train Qwen2.5-VL model with LoRA or full fine-tuning')
+    parser.add_argument('--use_lora', type=bool, default=True,
+                       help='Use LoRA (True) or full fine-tuning (False). Default: True')
+
+    args = parser.parse_args()
+    use_lora = args.use_lora == True
```
Comment (zhangfaen) [inline finetune_distributed.py]: use_lora = args.use_lora == True
直接改为：
use_lora = args.use_lora

Comment (zhangfaen) [inline finetune_distributed.py]: 这行不需要

File: finetune_distributed.py [hunk 3]
```diff
@@ -242,26 +481,64 @@ def train():
...
+    
+    train(use_lora=use_lora)
```
Comment (zhangfaen) [inline finetune_distributed.py]: train(use_lora=args.use_lora)

File: finetune_distributed.py [hunk 4]
```diff
@@ -185,7 +189,209 @@ def write_chat_template(processor, output_dir):
...
+        ],
+        bias="none",
+        use_rslora=False,
+        use_dora=False,
+    )
+    
+    return lora_config
+
+
+def apply_lora_to_model(model, logger=None):
```
Comment (zhangfaen) [inline finetune_distributed.py]: 不要 logger 这个参数。

File: finetune_distributed.py [hunk 5]
```diff
@@ -185,7 +189,209 @@ def write_chat_template(processor, output_dir):
...
+        logger: Logger instance (can be None in distributed training)
+        
+    Returns:
+        model: Model with LoRA applied
+    """
+    lora_config = setup_lora_config()
+    model = get_peft_model(model, lora_config)
+    
+    # Print trainable parameters info (only on main process)
+    if logger:
```
Comment (zhangfaen) [inline finetune_distributed.py]: 和其他地方一致的方式。

if accelerator.is_local_main_process:
  model.print_trainable_parameters()
  logger.info("Successfully applied LoRA to the model")

File: finetune_distributed.py [hunk 6]
```diff
@@ -185,7 +189,209 @@ def write_chat_template(processor, output_dir):
...
+        model.print_trainable_parameters()
+        logger.info("Successfully applied LoRA to the model")
+    
+    return model
+        
+    
+
+
+def plot_training_loss(loss_history, epoch_losses, output_dir, logger, training_type="LoRA"):
```
Comment (zhangfaen) [inline finetune_distributed.py]: 不要logger这个参数。 同上

File: finetune_distributed.py [hunk 7]
```diff
@@ -185,7 +189,209 @@ def write_chat_template(processor, output_dir):
...
+    json_path = os.path.join(output_dir, f'{training_type.lower()}_loss_data.json')
+    with open(json_path, 'w') as f:
+        json.dump(loss_data, f, indent=2)
+    
+    if logger:
+        logger.info(f"{training_type} loss plots saved to {plot_path}")
+        logger.info(f"{training_type} loss data saved to {json_path}")
+
+
+def compare_training_methods(lora_loss_data, full_loss_data, output_dir, logger):
```
Comment (zhangfaen) [inline finetune_distributed.py]: 不要logger这个参数。 同上

File: finetune_distributed.py [hunk 8]
```diff
@@ -199,6 +405,17 @@ def train():
         "Qwen/Qwen2.5-VL-3B-Instruct", torch_dtype="bfloat16"
     )
 
+    # Apply LoRA to the model based on parameter
+    if use_lora:
+        if accelerator.is_local_main_process:
+            logger.info("Applying LoRA to the model...")
+        # Pass logger only on main process, None on other processes
+        current_logger = logger if accelerator.is_local_main_process else None
```
Comment (zhangfaen) [inline finetune_distributed.py]: 不要这行。

因为 未来就算不是 main_process， 也可以使用 logger 记录一些日志信息等

**Review by zhangfaen — COMMENTED** [review]
整体很好。还有一些小的地方。

**Comment by pinecypressfxd** [comment]
有道理，简单的子函数内部增加logger不可控，主函数调用次数多的话尤其是多进程，冗余过多。我都修改了，重新pull request
